### PR TITLE
Jobs implementation

### PR DESCRIPTION
**Pull request by silvestrpredko**

### CODE REVIEW

File: beluga-aws-sdk/src/jobs/mod.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            )
+            .await?;
+
+        let accepted = format!("$aws/things/{}/jobs/get/accepted", self.mqtt.thing_name());
+        let rejected = format!("$aws/things/{}/jobs/get/rejected", self.mqtt.thing_name());
+
+        let mut subscriber = self
+            .mqtt
+            .subscribe_many([accepted.as_str(), rejected.as_str()], QoS::AtLeastOnce)
+            .await?;
```
Comment (scootermon) [inline beluga-aws-sdk/src/jobs/mod.rs:45]: we should subscribe before publishing to avoid a race condition.

Does the aws sdk also only temporarily subscribe to the accepted and rejected topics? I would've assumed the subscribes to these topics once at the start.

File: beluga-aws-sdk/src/jobs/datetime.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+use std::ops::Add;
+use std::time::Duration;
+
+use chrono::prelude::*;
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Deserializer, Serializer};
+
+pub(super) fn serialize<S>(date: &Option<DateTime<Utc>>, serializer: S) -> Result<S::Ok, S::Error>
```
Comment (scootermon) [inline beluga-aws-sdk/src/jobs/datetime.rs:8]: Can you add the unit tests from https://github.com/inomotech/iot-humpback/blob/890793e85576b08ef71656beedfaa7b085a27336/packages/hb-core/src/iot_core/client.rs#L146 as well?

File: beluga-aws-sdk/src/jobs/datetime.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+pub(super) fn deserialize<'de, D>(deserializer: D) -> Result<Option<DateTime<Utc>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Option::<f64>::deserialize(deserializer).map(|seconds_opt| {
+        seconds_opt.map(|seconds| {
+            DateTime::from_naive_utc_and_offset(
+                NaiveDateTime::UNIX_EPOCH.add(Duration::from_secs_f64(seconds)),
```
Comment (scootermon) [inline beluga-aws-sdk/src/jobs/datetime.rs:35]: I believe this will panic on overflow, right? Can we add a unit test for this case and make sure that we don't panic?

File: beluga-aws-sdk/src/jobs/mod.rs [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    pub async fn get(&mut self) -> Result<(Vec<Job>, Vec<Job>)> {
+        self.mqtt
+            .publish(
+                format!("$aws/things/{}/jobs/get", self.mqtt.thing_name()).as_str(),
+                QoS::AtLeastOnce,
+                false,
+                bytes::Bytes::copy_from_slice(&serde_json::to_vec(&GetPendingJobsExecutionReq {
+                    token: format!("get-jobs-{}", self.mqtt.thing_name()),
```
Comment (scootermon) [inline beluga-aws-sdk/src/jobs/mod.rs:49]: Shouldn't we pseudo-randomly generate this token (I guess we can just add a counter to it) so we can be sure we're reading the correct message (since the subscriber now buffers messages from the past).

File: beluga-aws-sdk/src/jobs/mod.rs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        let packet = self.subscriber.recv().await?;
+
+        if packet.topic == get_accepted(self.mqtt.thing_name()) {
+            let GetPendingJobsExecutionResp {
+                in_progress_jobs,
+                queued_jobs,
+                token,
+                ..
+            } = serde_json::from_slice::<GetPendingJobsExecutionResp>(&packet.payload)?;
+            if token.is_some_and(|token| token == format!("get-jobs-{}", self.mqtt.thing_name())) {
```
Comment (scootermon) [inline beluga-aws-sdk/src/jobs/mod.rs:63]: can we put this in a loop to basically say:

```rust
loop {
  let packet = self.subscriber.recv().await?;
  if !packet_matches_our_request() {
    continue;
  }
  // process job execution resp
}
```

This let's us gracefully deal with the case when there are still unrelated messages in the subscriber.

File: beluga-mqtt/src/lib.rs [hunk 6]
```diff
@@ -165,47 +180,24 @@ impl MqttClient {
     /// # Returns
     /// A [Result] containing a [Subscriber] if the subscriptions are
     /// successful, otherwise an [Error].
-    pub async fn subscribe_many(
+    pub async fn subscribe_many<'a>(
         &self,
-        topics: impl IntoIterator<Item = &str>,
+        topics: impl IntoIterator<Item = &'a str> + 'a,
```
Comment (scootermon) [inline beluga-mqtt/src/lib.rs:185]: We might as well make topics generic over `AsRef<str>`, right?
As in `where It: IntoIterator, It::Item: AsRef<str>`.
That way you don't need the `as_str()` when using `format!()`.

File: beluga-mqtt/src/manager.rs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    /// Returns a reference to the `Sender` for the given topic, if the
+    /// topic is currently subscribed.
+    pub(super) fn sender(&self, topic: &str) -> Option<&Sender<Publish>> {
+        self.subscribed.get(topic)
+    }
+
+    /// Returns a receiver for the given topic, if the topic is currently
+    /// subscribed
+    pub(super) fn receiver(&self, topic: &str) -> Option<Receiver<Publish>> {
+        self.subscribed.get(topic).map(|sender| sender.subscribe())
```
Comment (scootermon) [inline beluga-mqtt/src/manager.rs:22]: I think at this point we need to keep track of the number of active subscribers to a given topic (basically reference counting).

Right now if we subscribe to topic "a" twice and then one of the two subscribers calls "unsubscribe" it will unsubscribe both of them.

File: beluga-aws-sdk/src/jobs/mod.rs [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+                let in_progress = futures::stream::iter(in_progress_jobs.into_iter())
+                    .map(|summary| async { Job::from_summary(summary, self.mqtt.clone()).await })
+                    .buffer_unordered(10)
+                    .try_collect::<Vec<_>>()
+                    .await?;
+
+                let queued = futures::stream::iter(queued_jobs.into_iter())
+                    .map(|summary| async { Job::from_summary(summary, self.mqtt.clone()).await })
+                    .buffer_unordered(10)
```
Comment (scootermon) [inline beluga-aws-sdk/src/jobs/mod.rs:72]: I think it might be an issue if we pre-emptively subscribe to all topics related to every queued job. Let's say the pending queue contains like 200 pending jobs for instance.

I think we could either:

1. return a different "lazy job" type that needs to be upgrade into the real `Job` type.
2. return streams from this function rather than collecting the jobs into vecs eagerly.

What are your thoughts?

File: beluga-mqtt/src/manager.rs [hunk 9]
```diff
@@ -0,0 +1,86 @@
+use std::collections::{HashMap, HashSet};
+
+use rumqttc::Publish;
+use tokio::sync::broadcast::{self, Receiver, Sender};
+
+#[derive(Debug, Default)]
+pub(super) struct SubscriberManager {
```
Comment (scootermon) [inline beluga-mqtt/src/manager.rs:7]: This type definitely deserves unit tests especially surrounding the edge cases.

I see for instance that you already thought of the case where someone re-subscribes to a topic before the previous "unsubscribed" happens, but it would be great to have this interaction as a unit test as well.

File: beluga-mqtt/src/lib.rs [hunk 10]
```diff
@@ -240,35 +232,137 @@ impl MqttClient {
...
+    let removed_topics = futures::stream::iter(manager.scheduled())
+        .map(ToOwned::to_owned)
+        .map(|topic| async move {
+            client.unsubscribe(&topic).await?;
+            Ok::<String, rumqttc::ClientError>(topic)
+        })
+        .buffer_unordered(10)
+        .filter_map(|res| async { res.ok() })
+        .collect::<Vec<_>>()
+        .await;
```
Comment (scootermon) [inline beluga-mqtt/src/lib.rs:354]: Does this work right now? I assume it does but something seems off to me.
This `process_packet` function essentially blocks `event_loop.poll()` since it happens as part of the event handling.
Now the problem I'm seeing is that `event_loop.poll()` is what actually drives rumqtt's logic, so how can we unsubscribe here?

Is there something I'm missing?

File: beluga-aws-sdk/src/jobs/datetime.rs (hunk 3 above)
Comment (silvestrpredko) [reply beluga-aws-sdk/src/jobs/datetime.rs:35]: Yes, having tests is necessary since many places require testing. Let's create a separate task for it. For instance, it would be ideal to have a test broker for the MQTT client to test basic scenarios. What do you think?

File: beluga-aws-sdk/src/jobs/mod.rs (hunk 4 above)
Comment (silvestrpredko) [reply beluga-aws-sdk/src/jobs/mod.rs:49]: Yeah, agree.

File: beluga-mqtt/src/manager.rs (hunk 7 above)
Comment (silvestrpredko) [reply beluga-mqtt/src/manager.rs:22]: For now, you need to call unsubscribe on a client instance, subscriber only could receive a message. This question is more for an API design that we want at the end, let's discuss that.

File: beluga-aws-sdk/src/jobs/mod.rs (hunk 8 above)
Comment (silvestrpredko) [reply beluga-aws-sdk/src/jobs/mod.rs:72]: Hmmm, let's discuss that.

File: beluga-mqtt/src/lib.rs (hunk 10 above)
Comment (silvestrpredko) [reply beluga-mqtt/src/lib.rs:354]: From my debugging it puts messages in a queue, seems like it's completely safe to do these things in a poll.

Comment (scootermon) [reply beluga-mqtt/src/lib.rs:354]: @silvestrpredko I looked into the rumqttc source code. They're using flume for the underlying channel and `unsubscribe` calls [Sender::send_async](https://docs.rs/flume/latest/flume/struct.Sender.html#method.send_async), which will "block" when the channel is full.

Comment (silvestrpredko) [reply beluga-mqtt/src/lib.rs:354]: Yeah, that makes sense.

**Review by scootermon — APPROVED** [review]
